editor/src/messages/viewport/viewport_message_handler.rs

```diff
@@ -0,0 +1,391 @@
+			ViewportMessage::Trigger => {}
+		}
+
+		let physical_bounds = self.physical_bounds();
+		responses.add(FrontendMessage::UpdateViewportPhysicalBounds {
+			x: physical_bounds.x,
+			y: physical_bounds.y,
+			width: physical_bounds.width,
+			height: physical_bounds.height,
+		});
+
+		responses.add(NavigationMessage::CanvasPan { delta: DVec2::ZERO });
+		responses.add(NodeGraphMessage::SetGridAlignedEdges);
+
+		responses.add(DeferMessage::AfterGraphRun {
+			messages: vec![
+				DeferMessage::AfterGraphRun {
+					messages: vec![DeferMessage::TriggerNavigationReady.into()],
+				}
+				.into(),
+			],
+		});
+	}
+
+	advertise_actions!(ViewportMessageDiscriminant;);
+}
+
+impl ViewportMessageHandler {
+	pub fn scale(&self) -> f64 {
+		self.scale
+	}
+
+	pub fn logical_bounds(&self) -> LogicalBounds {
+		self.bounds
+	}
+
+	pub fn physical_bounds(&self) -> PhysicalBounds {
+		self.logical_bounds().to_physical(self.scale)
+	}
+
+	pub fn logical_offset(&self) -> LogicalPoint {
+		LogicalPoint { x: self.bounds.x, y: self.bounds.y }
+	}
+
+	pub fn physical_offset(&self) -> PhysicalPoint {
+		self.logical_offset().to_physical(self.scale)
+	}
+
+	pub fn logical_size(&self) -> LogicalPoint {
+		LogicalPoint {
+			x: self.bounds.width,
+			y: self.bounds.height,
+		}
+	}
+
+	pub fn physical_size(&self) -> PhysicalPoint {
+		self.logical_size().to_physical(self.scale)
+	}
+
+	pub fn logical_center_in_viewport_space(&self) -> LogicalPoint {
+		let logical_size = self.logical_size();
+		LogicalPoint {
+			x: logical_size.x / 2.0,
+			y: logical_size.y / 2.0,
+		}
+	}
+
+	pub fn physical_center_in_viewport_space(&self) -> PhysicalPoint {
+		self.logical_center_in_viewport_space().to_physical(self.scale)
+	}
+
+	pub fn logical_center_in_window_space(&self) -> LogicalPoint {
+		self.apply_offset_to_logical_point(self.logical_center_in_viewport_space())
+	}
+
+	pub fn physical_center_in_window_space(&self) -> PhysicalPoint {
+		self.logical_center_in_window_space().to_physical(self.scale)
+	}
+
+	pub fn is_in_logical_bounds<T: Into<LogicalPoint>>(&self, point: T) -> bool {
+		let point = point.into();
+		point.x >= self.bounds.x && point.y >= self.bounds.y && point.x <= self.bounds.x + self.bounds.width && point.y <= self.bounds.y + self.bounds.height
+	}
+
+	pub fn is_in_physical_bounds<T: Into<PhysicalPoint>>(&self, point: T) -> bool {
+		let point = self.convert_physical_to_logical_point(point.into());
+		self.is_in_logical_bounds(point)
+	}
+
+	pub fn convert_physical_to_logical(&self, physical: f64) -> f64 {
+		physical.to_logical(self.scale)
+	}
+
+	pub fn convert_logical_to_physical(&self, logical: f64) -> f64 {
+		logical.to_physical(self.scale)
+	}
+
+	pub fn convert_physical_to_logical_point<T: Into<PhysicalPoint>>(&self, physical: T) -> LogicalPoint {
+		physical.into().to_logical(self.scale)
+	}
+
+	pub fn convert_logical_to_physical_point<T: Into<LogicalPoint>>(&self, logical: T) -> PhysicalPoint {
+		logical.into().to_physical(self.scale)
+	}
```

Comment:
I'm wondering if it would make for a nicer API if we only returned logical points, but included the scale so you could call `.to_physical()` on it or use an into conversion. It does feel award that you basically always need the viewport to perform operations on points. The advantage is of course that this helps prevent using and outdated scale, but I'm not sure how often that would be applicable